app/src/main/java/com/vultisig/wallet/data/repositories/BlockChainSpecificRepository.kt

```diff
@@ -0,0 +1,96 @@
+package com.vultisig.wallet.data.repositories
+
+import com.vultisig.wallet.data.api.CosmosApiFactory
+import com.vultisig.wallet.data.api.EvmApiFactory
+import com.vultisig.wallet.data.api.MayaChainApi
+import com.vultisig.wallet.data.api.SolanaApi
+import com.vultisig.wallet.data.api.ThorChainApi
+import com.vultisig.wallet.data.models.TokenStandard
+import com.vultisig.wallet.data.models.TokenValue
+import com.vultisig.wallet.models.Chain
+import com.vultisig.wallet.presenter.keysign.BlockChainSpecific
+import java.math.BigInteger
+import javax.inject.Inject
+
+internal interface BlockChainSpecificRepository {
+
+    suspend fun getSpecific(
+        chain: Chain,
+        address: String,
+        gasFee: TokenValue,
+    ): BlockChainSpecific
+
+}
+
+internal class BlockChainSpecificRepositoryImpl @Inject constructor(
+    private val thorChainApi: ThorChainApi,
+    private val mayaChainApi: MayaChainApi,
+    private val evmApiFactory: EvmApiFactory,
+    private val solanaApi: SolanaApi,
+    private val cosmosApiFactory: CosmosApiFactory,
+) : BlockChainSpecificRepository {
+
+    override suspend fun getSpecific(
+        chain: Chain,
+        address: String,
+        gasFee: TokenValue,
+    ): BlockChainSpecific = when (chain.standard) {
+        TokenStandard.THORCHAIN -> {
+            val account = if (chain == Chain.mayaChain) {
+                mayaChainApi.getAccountNumber(address)
+            } else {
+                thorChainApi.getAccountNumber(address)
+            }
+
+            BlockChainSpecific.THORChain(
+                accountNumber = BigInteger(account.accountNumber),
+                sequence = BigInteger(account.sequence ?: "0"),
+            )
+        }
+
+        TokenStandard.EVM -> {
+            val evmApi = evmApiFactory.createEvmApi(chain)
+
+            val gasLimit = 23000 // TODO gas limit may differ for chains
```

Comment:
you can pass in coin , and coin.IsNativeToken , if it is true means it is ETH , otherwise it is ERC20 , so the gas limit should set accordingly